src/store/index-level.ts

```diff
@@ -216,46 +287,203 @@ export class IndexLevel {
+    const matches: IndexedItem<T>[] = [];
+
+    for await (const [ key, value ] of indexPartition.iterator(iteratorOptions, options)) {
       // if "greater-than" is specified, skip all keys that contains the exact value given in the "greater-than" condition
-      if ('gt' in rangeFilter && IndexLevel.extractValueFromKey(key) === this.encodeValue(rangeFilter.gt)) {
+      if ('gt' in rangeFilter && this.extractValueFromKey(key) === this.encodeValue(rangeFilter.gt)) {
         continue;
       }
 
       // immediately stop if we arrive at an index entry for a different property
-      if (!key.startsWith(propertyName)) {
+      if (!key.startsWith(matchPrefix)) {
         break;
       }
 
-      matches.push(dataId);
+      const itemId = this.extractItemId(key);
+      matches.push({ id: itemId, value: JSON.parse(value) });
     }
 
+    // we gather the lte matches separately to include before or after the results depending on the sort.
+    const lteMatches:IndexedItem<T>[] = [];
     if ('lte' in rangeFilter) {
       // When `lte` is used, we must also query the exact match explicitly because the exact match will not be included in the iterator above.
-      // This is due to the extra data (CID) appended to the (property + value) key prefix, e.g.
-      // key = 'dateCreated\u0000"2023-05-25T11:22:33.000000Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // the value would be considered greater than { lte: `dateCreated\u0000"2023-05-25T11:22:33.000000Z"` } used in the iterator options,
+      // This is due to the extra data appended to the (property + value) key prefix, e.g.
+      // the key 'watermark\u0000dateCreated\u0000"2023-05-25T11:22:33.000000Z"\u000001HBY2E1TPY1W95SE0PEG2AM96\u0000bayfreigu....'
+      // would be considered greater than { lte: 'watermark\u0000dateCreated\u0000"2023-05-25T11:22:33.000000Z"` } used in the iterator options,
       // thus would not be included in the iterator even though we'd like it to be.
-      for (const dataId of await this.findExactMatches(tenant, propertyName, rangeFilter.lte, options)) {
-        matches.push(dataId);
+      //
+      // we also only include the cursor ONLY if it is relevant to the exact property in the 'lte' filter.
+      const lteCursor = cursor && this.extractValueFromKey(cursor) === this.encodeValue(rangeFilter.lte) ? cursor : undefined;
+      for (const item of await this.findExactMatches(tenant, propertyName, rangeFilter.lte, sortProperty, sortDirection, lteCursor, options)) {
+        lteMatches.push(item);
       }
     }
 
-    return matches;
+    // if the iterator is reversed and the results should be in Ascending order, we reverse the iterated matches.
+    // if the iterator is not reversed and the results should be in Descending order, we also reverse the iterated matches.
+    if ((iteratorOptions.reverse === true && sortDirection === SortOrder.Ascending) ||
```

Comment:
Sanity: I am failing to understand the point of sorting here and in all other "match" methods. Don't they get put into `matches` later which is a `Map` thus all this work is throw-away?